### ISSUE TITLE
New command line option `--orig` to rename conflicting files

### ISSUE THREAD

**Issue by lassik** [aspiers/stow#4]
Hello!

Thank you for stow -- I'm enjoying it a lot. It has become the first tool that I install on a new machine after git. Also, I hope github is an OK channel for sending feature requests.

Here's my request: It'd be nice to have a new command line option `--orig` (or similar) to make backups of conflicting files in the target directory instead of failing on conflict. For example:

* `hosts` => `hosts.orig`
* `ntp.conf` => `ntp.conf.orig`
* `.bashrc` => `.bashrc.orig`

I often use stow to link files from a git repo into `$HOME` or `/etc`, and Linux distributions ship with default configuration files in these locations that stow will trip over. The `--adopt` option does not solve my problem because I don't want to copy the OS default configs into my git repo - I want to do the opposite and overwrite the OS default configs (but keep backups of them).

Leaving each backup file in the same directory as the original and using a constant backup extension (`.orig`) has the nice effect that one can do e.g. `find /etc -name "*.orig"` to find all overwritten configs. Likewise, it's fairly easy to run diff to see changes from the OS defaults.

Here's a demonstration of what currently happens:

    $ mkdir demo
    $ mkdir -p demo/stow/test
    $ touch demo/foo demo/stow/test/foo
    $ tree demo --charset ascii
    demo
    |-- foo
    `-- stow
        `-- test
            `-- foo

    2 directories, 2 files
    $ (cd demo/stow && stow test)
    WARNING! stowing test would cause conflicts:
      * existing target is neither a link nor a directory: foo
    All operations aborted.

This is what I would like to have:

    $ (cd demo/stow && stow --orig -vv test)
    ...
    RENAME: foo => foo.orig
    LINK: foo => stow/test/foo
    ...
    $ tree demo --charset ascii
    demo
    |-- foo -> stow/test/foo
    |-- foo.orig
    `-- stow
        `-- test
            `-- foo

    2 directories, 3 files

Here's a succint description of the behavior that `--orig` would have, covering edge cases:
* IF dst already exists THEN
  * IF dst.orig already exists THEN give up and stop with an error
  * ELSE rename dst to dst.orig(even if dst is a symlink, a directory, etc.)
    * IF the rename fails due to permissions etc. THEN give up and stop with an error

The option could also be named `--backup` or similar, but I personally like `--orig` because it's short and mnemonic (it reminds the user that the backup filename will have the extension `.orig`, which is also a de facto standard extension for such files).

Would you accept a feature like this into stow?

Kind regards,
Lassi

**Comment by lassik** [comment]
More sketching:

What if we instead moved the backup files into a designated backup directory that mirrors directory structure in the same way that stow packages do. E.g.

    $ tree -a --charset ascii
    .
    |-- .git
    |-- .gitignore
    `-- my-configs
        `-- etc
            `-- ntp.conf
    $ cat .gitignore
    /defaults/
    $ stow my-configs -vvt / --backupdir defaults
    ...
    MKDIR: defaults       # because it didn't exist already
    MKDIR: defaults/etc   # because it didn't exist already
    COPY: /etc/ntp.conf => defaults/etc/ntp.conf  # or this could be a rename...
    REMOVE: /etc/ntp.conf         # ...if file system boundaries are not crossed
    LINK: etc/ntp.conf => /etc/ntp.conf
    ...
    $ tree -a --charset ascii
    .
    |-- defaults
    |   `-- etc
    |       `-- ntp.conf
    |-- .git
    |-- .gitignore
    `-- my-configs
        `-- etc
            `-- ntp.conf

If the user wanted to, they could also append a filename extension of their choice to all backup files. If they didn't give such an extension, nothing would be appended. E.g.:

    $ stow my-configs -vvt / --backupext .orig
    ...
    RENAME: /etc/ntp.conf => /etc/ntp.conf.orig
    LINK: etc/ntp.conf => /etc/ntp.conf
    ...

It would also be possible to combine `--backupdir` and `--backupext` in a natural way:

    $ stow my-configs -vvt / --backupdir defaults --backupext .orig
    ...
    COPY: /etc/ntp.conf => defaults/etc/ntp.conf.orig
    REMOVE: /etc/ntp.conf
    LINK: etc/ntp.conf => /etc/ntp.conf
    ...

**Comment by Deichscheich** [comment]
Great work in outlining this, lassik! This feature would be reeeeally useful.

**Comment by SyntaxColoring** [comment]
:+1:

**Comment by aspiers** [comment]
Yes thanks, this is a good feature request.  Please look at [`patch(1)`](http://linux.die.net/man/1/patch) for ideas on how to define options which control the algorithm for choosing where to back up to.

**Comment by lassik** [comment]
Thanks for the interest!

Good tip on `patch(1)`. The above proposed `--backupdir` and `--backupext` seem equivalent to `patch`'s `--prefix` and `--suffix`, respectively. (Except that `patch`'s `--prefix` need not be a directory if it doesn't end in a slash.)

`patch`'s naming scheme support seems very comprehensive with the `--basename-prefix` and `--version-control` options (numbered backup files, etc.) in addition to the above. Do you think anyone would find these useful in `stow`'s context? My personal intuition is that they would be over the top.

**Comment by SyntaxColoring** [comment]
Yeah, that seems like overkill.

**Comment by sebastienbarre** [comment]
This would be a nice feature. Did it make it in?

**Comment by aspiers** [comment]
Nope, someone has to code it first ;-)

**Comment by astier** [comment]
Would be really cool if it would be possible to overwrite existing files.

**Comment by H3mul** [comment]
If someone badly needs this (like I do!) and can't wait for this to be merged into upstream, and is on Arch, you're welcome to my [PKGBUILD](https://raw.githubusercontent.com/H3mul/stow/dev/PKGBUILD) in the meantime.
